ext/random/random.c

```diff
@@ -503,14 +504,30 @@ PHPAPI int php_random_bytes(void *bytes, size_t size, bool should_throw)
 		}
 		return FAILURE;
 	}
-#elif HAVE_DECL_ARC4RANDOM_BUF && ((defined(__OpenBSD__) && OpenBSD >= 201405) || (defined(__NetBSD__) && __NetBSD_Version__ >= 700000001) || defined(__APPLE__) || defined(__GLIBC__))
+#elif HAVE_DECL_ARC4RANDOM_BUF && ((defined(__OpenBSD__) && OpenBSD >= 201405) || (defined(__NetBSD__) && __NetBSD_Version__ >= 700000001 && __NetBSD_Version__ < 1000000000) || \
+  defined(__APPLE__) || defined(__GLIBC__))
+	/*
+	 * OpenBSD until there is a valid equivalent
+	 * or NetBSD before the 10.x release
+	 * falls back to arc4random_buf
+	 * giving a decent output but also
+	 * being implemented in the userland only
+	 * does not necessarily provide the highest
+	 * performance.
+	 * Older macOs releases fall also into this
+	 * category for reasons explained above.
+	 */
 	arc4random_buf(bytes, size);
 #else
 	size_t read_bytes = 0;
 	ssize_t n;
-# if (defined(__linux__) && defined(SYS_getrandom)) || (defined(__FreeBSD__) && __FreeBSD_version >= 1200000) || (defined(__DragonFly__) && __DragonFly_version >= 500700) || defined(__sun)
-	/* Linux getrandom(2) syscall or FreeBSD/DragonFlyBSD getrandom(2) function*/
-	/* Keep reading until we get enough entropy */
+# if (defined(__linux__) && defined(SYS_getrandom)) || (defined(__FreeBSD__) && __FreeBSD_version >= 1200000) || (defined(__DragonFly__) && __DragonFly_version >= 500700) || \
+  defined(__sun) || (defined(__NetBSD__) && __NetBSD_Version__ >= 1000000000)
+	/* Linux getrandom(2) syscall or FreeBSD/DragonFlyBSD/NetBSD getrandom(2) function
+	 * Being a syscall, implemented in the kernel, getrandom offers higher performance
+	 * compared to the arc4random api albeit a fallback to /dev/urandom is considered.
```

Comment:
Is this actually true? My understanding is that a syscall is generally slower than a userland function call, due to the necessary context switch. It should be more secure, though.

In any case it certainly contradicts the intent of #8984, which specifically switched to arc4random_buf() for glibc in an attempt to improve performance (which ultimately didn't actually do anything, because glibc's final implementation to my understanding is just a thin wrapper around getrandom(2)).

Note that for the CSPRNG I would generally favor security over speed, it should be plenty fast for almost any use case anyway, so I'm inclined to take this change that switched NetBSD to the syscall and to revert #8984 to unify all the Unices to getrandom (if available).

Reply:
ah you re right 🤦🏻‍♂️ thanks for pointing out, indeed even with VDSO it s not faster.